src/librustc_mir/transform/const_prop.rs

```diff
@@ -915,6 +916,29 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
             }
         } else {
             match statement.kind {
+                StatementKind::SetDiscriminant { ref place, .. } => {
+                    match self.can_const_prop[place.local] {
+                        ConstPropMode::FullConstProp | ConstPropMode::OnlyInsideOwnBlock => {
+                            if self.use_ecx(|this| this.ecx.statement(statement)).is_some() {
+                                trace!(
+                                    "propped discriminant into {:?}",
+                                    place,
+                                );
+                                trace!(
+                                    "propped into local {:?} which is restricted to its block. \
+                                    Will remove it from const-prop after block is finished.",
+                                    place.local
+                                );
+                                self.locals_of_current_block.insert(place.local);
```

Comment:
Don't we only want this for `ConstPropMode::OnlyInsideOwnBlock`?

Reply:
`SetDiscriminant` and `Assign` are indistinguishable. If there are no mutations other than a single `SetDiscriminant` in the entire `mir::Body`, we can safely do this, just like we can do an assignment if there are no mutations but that single assignment.

Reply:
You get such a situation e.g. for `let x: Option<SomeThing> = None;` This translates to a single `SetDiscriminant` in MIR without any other assignments to `x`.

Reply:
Right but if this local is `ConstPropMode::FullConstProp`, then we don't want to schedule it for removal at the end of this block. But I guess this doesn't really matter because #73757 removes `locals_of_current_block` anyway.

Reply:
ooh. Yea, I totally forgot about that. I'm just going to wait for #73757 and rebase over that.